### PR TITLE
scripts: ci: check_compliance.py: Add clang-format check

### PR DESCRIPTION
**Pull request by pdgendt**
Add a new compliance check that reports any clang-format issues on the git diff.

See [the test PR](https://github.com/zephyrproject-rtos/zephyr/pull/75117/files) on how these are reported.

### CODE REVIEW

**Review by gmarull — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
At the moment clang-format adoption is pretty much a mix of reviewer/author preference, blanket adoption on all diffs does not strike me as a very good idea, you'd end up with a permanently half-compliant code base. IMO if we are doing this it would have to be per file/area and on the whole file.

**Comment by pdgendt** [comment]
I thought I'd casually submit this PR and watch the chaos unfold...

Joking aside, starting with the diffs is the way to go IMO, I'm usually one of those irritating reviewers with a lot of "nit" comments, which can be easily captured with a check.

**Comment by fabiobaltieri** [comment]
> Joking aside, starting with the diffs is the way to go IMO

It's really not something that would work well now, take the column limit for example: most files are still on 80, clang-format really wants it to be 100. In this state you'd at least try to have the file using something somewhat coherent, if you enable this you'd end up forcing boclks on 100 on anything that is recently edited.

**Comment by pdgendt** [comment]
Reading up on #48785, #52712, #73465 and probably more...

Maybe we could turn the compliance check into a warning instead of an error? This way reviewers can refer to it, but a submitter can also do updates based on the result before that.

**Comment by fabiobaltieri** [comment]
> Maybe we could turn the compliance check into a warning instead of an error?

That's another issue, AFAIK you can't just have an "informational" check in GitHub, either the check fails or not. Though I've seen annotations that looks like warning in the code. Can it be done? Would be nice to have checks that are more like a note for the reviewer and don't need org superpowah to merge.

**Comment by pdgendt** [comment]
> > Maybe we could turn the compliance check into a warning instead of an error?
> 
> That's another issue, AFAIK you can't just have an "informational" check in GitHub, either the check fails or not. Though I've seen annotations that looks like warning in the code. Can it be done? Would be nice to have checks that are more like a note for the reviewer and don't need org superpowah to merge.

I already modified the workflow to have an annotation without failing the action.

**Comment by keith-zephyr** [comment]
It's my preference that we have clang-format (or some other formatter) be the source of truth for code style.  This let's reviewers focus on the actual implementation and avoids any disagreement about style.  As @nashif mentioned, "no one argues with a bot".

However, there are some strong opinions against clang-format in the project.  So the current guidelines state that the documented code-style takes precedence over clang-format output if there is a mismatch.

**Comment by pdgendt** [comment]
@fabiobaltieri I've update the PR to add a "notice" annotation to the file:

image

See [the test PR](https://github.com/zephyrproject-rtos/zephyr/pull/75117/files#diff-8636dbd224f587a7993ec2655cabce4185a0aea2cb53d8065639d22ecd7cc4b7)

**Review by gmarull — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri I've updated the PR to add a "notice" annotation to the file:

Cool, still it's unclear what the endgame is but I don't necessarily hate this. :-)

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by pdgendt** [comment]
> Cool, still it's unclear what the endgame is but I don't necessarily hate this. :-)

If a reviewer agrees with clang-format, he can simply comment with "Fix the formatting, please". Instead of pointing out all the nits that he would like to see fixed.

File: scripts/ci/check_compliance.py
Comment (kartben) [inline scripts/ci/check_compliance.py:293]: the position (line #) of the notice looks weird in the test PR you shared? I think this should be adjusted as it's awkward to get the notice placed at a line that, in this particular example, is not even problematic.

**Review by kartben — CHANGES_REQUESTED** [review]

File: scripts/ci/check_compliance.py
Comment (pdgendt) [reply scripts/ci/check_compliance.py:293]: It's on the end of the hunk, which I think makes sense? Where would you put it otherwise?

The image in my previous comment is of by one as this was the previous version, the test PR looks good IMO.

Comment (pdgendt) [reply scripts/ci/check_compliance.py:293]: Annotations can be set to more lines using `endLine`, will update.

File: scripts/ci/check_compliance.py
Comment (kartben) [inline scripts/ci/check_compliance.py:295]: I was thinking something along these lines (LOL)?  
```suggestion
                for patch in patchset:
                    for hunk in patch:
                        for line in hunk:
                            # Set the annotation to the first offending line in the hunk
                            if line.is_removed:
                                self.fmtd_failure("notice", "ClangFormat", file, line=line.source_line_no,
                                                desc=f'\r\n{hunk}')
                                break
```

File: scripts/ci/check_compliance.py
Comment (kartben) [reply scripts/ci/check_compliance.py:293]: > Annotations can be set to more lines using `endLine`, will update.

ya, that'd help, I think. I think my main concern is that if it's a *line* annotation the actual position should somewhat make sense and directly point at something actionable

Comment (pdgendt) [reply scripts/ci/check_compliance.py]: Update, the start line is now printed in the workflow, but the annotation does appear to be the same as before.

**Review by simhein — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by keith-zephyr** [comment]
This is a good addition even if it only used to catch missing braces around conditionals.  I see that get missed in reviews often, and there isn't any existing compliance check for braces.

File: scripts/ci/check_compliance.py
Comment (keith-zephyr) [inline scripts/ci/check_compliance.py:1]: Nit for the commit message for the middle commit: "Allow to set add information about...".   Maybe you wanted "Allow setting information about the end line and end column. for a check failure".

Comment (pdgendt) [reply scripts/ci/check_compliance.py:1]: Updated

File: scripts/ci/check_compliance.py
Comment (kartben) [inline scripts/ci/check_compliance.py:300]: Sorry but I still think the notice should appear at the location of the first offending line, especially if you are to make the "diff" appear as part of the message, since you want people to be able to easily match the diff to what's below/after the notice, not above (IMO).

Please see https://github.com/zephyrproject-rtos/zephyr/pull/75165/files also for suggestions on how to make the notice even more useful, namely:

- "ClangFormat" is a bit too succinct of a title, IMO
- Outputting the hunk "header" is unnecessary, and it might make sense to align the diff at the beginning of the annotation

BEFORE:

image


AFTER:

image

**Review by kartben — CHANGES_REQUESTED** [review]

File: scripts/ci/check_compliance.py
Comment (pdgendt) [reply scripts/ci/check_compliance.py]: I've updated the PR, but simplified the handling to skip the first and last 3 lines, as these were always added surrounding the hunk.

**Review by kartben — DISMISSED** [review]
Thanks, @pdgendt!

**Review by keith-zephyr — DISMISSED** [review]

**Comment by nashif** [comment]
setting this for 4.0, need to see this more tested, having a new compliance check appear while we are stablizing and depending on how much noise it will create is not great. 
I also do not like the fact it does not say what the issue it is fixing.

Also need @jfischer-no  to give his opinion on this :)

**Comment by pdgendt** [comment]
> setting this for 4.0, need to see this more tested, having a new compliance check appear while we are stablizing and depending on how much noise it will create is not great.

It needs to be in CI to get more testing ground, it doesn't block anything, I would argue that a feature freeze is actually ideal as other PRs remain open for a longer amount of time, which we can use here.

> I also do not like the fact it does not say what the issue it is fixing.

You mean in the PR description?

> Also need @jfischer-no to give his opinion on this :)

Sure

**Comment by nashif** [comment]
> You mean in the PR description?

no, the comments added when compliance check is run

File: scripts/ci/check_compliance.py
Comment (nashif) [reply scripts/ci/check_compliance.py]: Well, I got confused with that new way and I think the BEFORE looks much better, becase the second violation:

image

 seems completely disconnected from the bot comment

**Comment by nashif** [comment]
> > setting this for 4.0, need to see this more tested, having a new compliance check appear while we are stablizing and depending on how much noise it will create is not great.
> 
> It needs to be in CI to get more testing ground, it doesn't block anything, I would argue that a feature freeze is actually ideal as other PRs remain open for a longer amount of time, which we can use here.

well, many PRs in the past have been blocked and delayed because of changes done using clang-format, even if the PR will not fail, if someone start applying suggestions that are not accepted just because of style changes at this stage, this will not be great for 3.7 and just a distraction.

**Comment by aescolar** [comment]
Adding DNM only because the PR still seems to have some ongoing discussion about the overall idea but it meets the merge criteria so we avoid accidentally merging it.

**Comment by pdgendt** [comment]
> Adding DNM only because the PR still seems to have some ongoing discussion about the overall idea but it meets the merge criteria so we avoid accidentally merging it.

@aescolar should we add it as a topic for the process working group?

**Comment by aescolar** [comment]
> @aescolar should we add it as a topic for the process working group?

I think it would make sense

**Comment by keith-zephyr** [comment]
Consensus from Process WG
* Want to merge this PR and make improvements if it proves too noisy or encourages the wrong behavior from PR authors.
* @nashif - please comment if you have objections to merging.

Process WG notes

@pdgendt - doesn't propose making this a block CI check, only informational
@jfischer-no - will this complain on everyline?
@pdgendt - only flags changed lines, and merges multiple reports together.  Tooling flags the first line with the problem.
@pdgendt - The network stack has a 80 column limit, so the bot may be too noisy.
@dleach02 - Can this be suppressed.
@pdgendt - this might require another rule to override the tools
@jfischer-no USB allows >80, but prefers shorter lines
@jfischer-no - Doesn't want to override preferences with clang--formatting
@fabiobaltieri - this tooling is providing a guideline
@jfischer-no - reducing the line limit to 90 would be preferred.
@pdgendt - can't change the bot to use different rules than what specified by .clang-format
@pdgendt - in favor of automation, but spamming of the bot could be problem.
@fabiobaltieri - generally clang-format doesn't handle macros well.  Suggest using clang-format off/on blocks.
@fabiobaltieri - this is half solution.  Internally we enforce clang-format - and initialily reviewers took clang-format without evaluation.  But now reviewers tell authors to use off/on blocks
@fabiobaltieri - can we mark files/areas that require clang format compliance
@jfischer-no - Linux does have different format files for different areas.
@fabiobaltieri - only sees one .clang-format for full project.  But Linux doesn't have any compliance checks.
@fabiobaltieri - compliance is enforce culturally by the maintainer of the area
@jfischer-no - how does Google use the clang-format check.  @fabiobaltieri - our CI runs clang--format and rejects the patch.
@fabiobaltieri - this approach required scrubbing entire repo to make all files compliant.
@keith-zephyr can the compliance be run on only certain areas of the project.
@pdgendt - This is probably possible with some additional scripting.
@dleach02 - This will cause confusion
@pdgendt  - some rules should be blockers, but clang-format doesn't support levels.  Missing curly braces is a common item that is missed.
@MaureenHelm - would like checks for for braces that is blocker.
@jfischer-no - suggests realigning back to Linux coding style and eliminate exceptions.
@keith-zephyr - misra requires the braces on all conditionals
@pdgendt - does see efforts during release cycles to fix misra violations
@fabiobaltieri - in favor trying the patch out and seeing how noisy or if it encourages people to clang-format entire files.
@simhein also agrees that trying out the CI.  One concern is that this can confuse new comers if the maintainer doesn't follow all the clang format rules (line length, etc).
@jfischer-no - can a label be used to run this optional?  @fabiobaltieri that would apply to the full PR, and it better to filter at file/directory level
@fabiobaltieri - in favor of merging as is and iterating to make improvements

**Review by gmarull — DISMISSED** [review]

**Comment by aescolar** [comment]
@pdgendt @keith-zephyr feel free to remove the DNM if you think it should be merged already

**Review by nashif — CHANGES_REQUESTED** [review]
I have a concern with the grouping of inline comments, it is VERY confusing, see https://github.com/zephyrproject-rtos/zephyr/pull/75117/files#r1659260744.

Fine with this being merged and improved over time.

**Comment by pdgendt** [comment]
@kartben you suggested to mark the first line instead of the last, what to do here?

**Comment by kartben** [comment]
> @kartben you suggested to mark the first line instead of the last, what to do here?

I'm not sure, if I remember correctly the is no (easy?) way to put an annotation exactly in front of each non-compliant line since we're dealing with chunks, so my guess is that it will always be confusing on way or the other based on which direction folks end up scrolling (and, maybe more importantly how large the chunk is)? I.e in https://github.com/zephyrproject-rtos/zephyr/pull/75117/files#r1659260744 I am pretty sure showing the annotation near the last line of the chunk would have had @nashif comment "was it supposed to catch this?" on the first line :)

Or another way to answer would be: @nashif do you have a suggestion on  what would be *not* confusing, to you at least?

I'm +1 on merging as @nashif said, as what really matter is that the "diff" with recommended formatted output is there, albeit not always where people expect it.

**Comment by nashif** [comment]
as I already commented here https://github.com/zephyrproject-rtos/zephyr/pull/75089#discussion_r1659408237, the way it was before makes more sense.

**Comment by pdgendt** [comment]
> > Or another way to answer would be: @nashif do you have a suggestion on what would be not confusing, to you at least?
> 
> as I already commented here [#75089 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/75089#discussion_r1659408237), the way it was before makes more sense.

I tend to agree, but again not a strong opinion here, I can change it to the previous way.

**Comment by pdgendt** [comment]
@nashif @kartben I've reverted back to putting the annotation at the end, is this a good start to get something in and evaluate?

**Review by gmarull — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: .github/workflows/compliance.yml
Comment (kartben) [inline .github/workflows/compliance.yml:100]: @pdgendt could you look at making sure the clangformat check only runs on actual C and header files? see https://github.com/zephyrproject-rtos/zephyr/pull/77093/files#annotation_25086099623

Comment (pdgendt) [reply .github/workflows/compliance.yml:100]: Created #77095

**Pull request merged**
